Backend/FcTypeChecker.hs

```diff
@@ -352,23 +367,38 @@ matchClause dc (u:us) qs def = do
   return (FcAltFc (FcConPat dc us') fc_rhs)
 matchClause _   []    _  _   = panic "matchClause: empty variables"
 
--- | Match equations according to (new) or rule
-matchOr :: [FcTmVar] -> [PmEqn] -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
-matchOr us ((((FcOrPat p1 p2):ps), rhs):qs) def = match us (((p1:ps), rhs) : ((p2:ps), rhs) : qs) def
-matchOr _ _ _ = panic ("matchOr: called on a non or-pattern equation")
-
--- | Match a list of equations according to variable, constructor or or-pattern rule
-matchVarConOr :: [FcTmVar] -> [PmEqn] -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
-matchVarConOr us (q@(((FcConPatNs _ _):_), _):qs) def = matchCon us (q:qs) def
-matchVarConOr us (q@(((FcVarPat   _  ):_), _):qs) def = matchVar us (q:qs) def
-matchVarConOr us (q@(((FcOrPat    _ _):_), _):qs) def = matchOr  us (q:qs) def
-matchVarConOr _  qs                                _   = panic ("matchVarCon: invalid equations: " ++ render (ppr qs))
+-- | Match an or-pattern
+matchOr :: [FcTmVar] -> ([PmEqn], PmEqn, [PmEqn]) -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
+matchOr (u:us) (preq, (((FcOrPat p1 p2):ps), rhs), postq) def = do
+  exp_ty      <- lookupTmVarM u                         -- Get the expected type of the or-pattern
+  tmVarTys    <- extractTmVarTys exp_ty (FcOrPat p1 p2) -- Extract all term variables + types inside the or-pattern
+  (tm, tm_ty) <- extendCtxTmZipM tmVarTys (match us [(ps, rhs)] def >>= tcTerm) -- Call match on the RHS of the or-pattern + type check
+  f           <- foldrM abstractTmVarTy tm tmVarTys     -- Create the abstraction term that will be put inside the let
+  let f_ty    = foldr (\(_,ty1) ty2 -> mkFcArrowTy ty1 ty2) tm_ty tmVarTys -- Calculate the type of the abstraction term
+  y           <- makeVar                                -- Create a fresh variable (this will be the let-bound variable)
+  let y_app   = foldl applyTmVarTy (FcTmVar y) tmVarTys -- Create an application term to be used as the "new right hand side"
+  let g_app   = FcGuarded [] y_app                      -- Wrap the application term in a Guarded right hand side
+  -- Call match on "everything else" by replacing the or pattern in-place with two new equations containing the
+  -- the two patterns and using the guarded application term as right hand sides.
+  matched     <- extendCtxTmM y f_ty (match (u:us) (preq ++ [(p1:ps, [g_app]), (p2:ps, [g_app])] ++ postq) def)
+  -- Return a let term, binding the fresh variable to the abstraction term and using the result of the above match as body
+  return (FcTmLet y f_ty f matched)
```

Comment:
I think I now understand the code better, but not really because of the comments :stuck_out_tongue: In general, if there is some intricate strategy, you should not use inline comments to describe it, but instead explain it all before the function definition (or somewhere else; definitely not interleaved with the code). After all, obvious comments like
   ```haskell
   y <- makeVar  -- Create a fresh variable
   ```
   are only making things worse.

I left a suggestion below (have not tried to typecheck or run this, I just edited to make things a little more clear).

If I understand correctly what your algorithm does, it compiles the checks after the or patterns twice, once when checking the whole match, and once more in the compiled RHS. Though this makes sense, it makes me wonder whether duplicating the whole equation would be actually cheaper. I mean to simply replace
```haskell
preq ++ [(((FcOrPat p1 p2):ps), rhs)] ++ postq
```
with
```haskell
preq ++ [(p1:ps, rhs), (p2:ps, rhs)] ++ postq
```
After all, how bad can this be compared to the duplication you introduce?


```suggestion
-- | Match an or-pattern.
matchOr :: [FcTmVar] -> ([PmEqn], PmEqn, [PmEqn]) -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
matchOr (u:us) (preq, (((FcOrPat p1 p2):ps), rhs), postq) def = do

  -- 1. Compute the pattern variables (and their types) that
  -- are shared between the two sides of the or-pattern.
  tmVarTys <- do
    exp_ty <- lookupTmVarM u     
    extractTmVarTys exp_ty (FcOrPat p1 p2) 
  
  -- 2. Compile and typecheck the rest of the equation (without the
  -- or-pattern) as a single match.
  (tm, tm_ty) <- extendCtxTmZipM tmVarTys $
                   match us [(ps, rhs)] def >>= tcTerm

  -- 3. Create a let-binding, capturing what step 2 computed, where the
  -- variables bound by the or-patterns are abstracted over.
  letf <- do
    ftm <- foldrM abstractTmVarTy tm tmVarTys
    let fty = mkFcArrowTys (map snd tmVarTys) tm_ty
    f <- makeVar
    return (FcTmLet f fty ftm)

  let grhs = FcGuarded [] $ foldl applyTmVarTy (FcTmVar f) tmVarTys
  letf <$> extendCtxTmM f fty (match (u:us) (preq ++ [(p1:ps, [grhs]), (p2:ps, [grhs])] ++ postq) def)
  where
    -- GEORGE: I'd argue that this one should be placed where `mkFcArrowTy` is
    -- currently placed.
    mkFcArrowTys :: [FcType] -> FcType -> FcType
    mkFcArrowTys tys ty = foldr (\ty1 ty2 -> mkFcArrowTy ty1 ty2) ty tys

```

Reply:
Yeah I think your style of documentation is way better haha. Will take a look tomorrow. But I don't understand what you mean with this:

>>> If I understand correctly what your algorithm does, it compiles the checks after the or patterns twice, once when checking the whole match, and once more in the compiled RHS. Though this makes sense, it makes me wonder whether duplicating the whole equation would be actually cheaper. I mean to simply replace ...

I only compile the remaining patterns and rhs of or-pattern equation once:
`(tm, tm_ty) <- extendCtxTmZipM tmVarTys (match us [(ps, rhs)] def >>= tcTerm)`

The final match is performed on information that is disjoint from the first match. (First match is only on the or-pattern equation and nothing else. The final match is on everything outside the or-pattern equation + two new equations containing only the pattern and application term). So there is no overlapping computation in the match (as far as I can see) and so no duplication.

Reply:
Maybe this image helps? 😛 
![IMG_0875](https://user-images.githubusercontent.com/11155505/79082111-9d9f1f80-7d23-11ea-864f-e8a0b10e11cf.jpg)

Reply:
> I only compile the remaining patterns and rhs of or-pattern equation once:
> `(tm, tm_ty) <- extendCtxTmZipM tmVarTys (match us [(ps, rhs)] def >>= tcTerm)`

Not really, `ps` ends up being an argument of function `match` in two/three separate places:
```haskell
  (tm, tm_ty) <- extendCtxTmZipM tmVarTys (match us [(ps, rhs)] def >>= tcTerm) -- Call match on the RHS of the or-pattern + type check
```
and
```haskell
  matched     <- extendCtxTmM y f_ty (match (u:us) (preq ++ [(p1:ps, [g_app]), (p2:ps, [g_app])] ++ postq) def)
```
No?

Reply:
Ah I see what you mean. There is indeed duplication of those `ps`. I've found two possible solutions. But I think some duplication is inevitable.

Reply:
Right, I think so too (that some duplication is inevitable). My point I guess was that instead of doing something so convoluted, it might be even better (performance-wise) to just replicate the whole equation (just replace the single equation `(((FcOrPat p1 p2):ps), rhs)` with two equations `[((p1:ps), rhs),  ((p2:ps), rhs)]`).

Reply:
But then you just have normal flattening without any let bindings? Or what exactly do you mean with `rhs` in this example?

**Option 1:** Call first match on (ps, rhs) and the second match on [([p1], g_app), ([p2], g_app)]
Prefer this option, because less duplication, however, not possible, because then we need to call the second match on a matrix where there is no longer the same amount of patterns in each row. This is problematic because then we either get the situation where:
- There is an equation / row, that has no patterns left, but there are term variables left in the list. The partitioning will then fail because it can’t match on a patternless equation
- There is an equation / row with patterns left, but there are no term variables left to bind them and so information is lost or we get binding errors.

**Option 2:** Call first match on ([], rhs) and second match on [((p1: ps), g_app), ((p2:ps), g_app)]. Still some duplication, but better than nothing.

Reply:
Indeed, you have normal flattening in what I describe, which is basically this:
```haskell
-- | Match an or-pattern
matchOr :: [FcTmVar] -> ([PmEqn], PmEqn, [PmEqn]) -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
matchOr (u:us) (preq, (((FcOrPat p1 p2):ps), rhs), postq) def =
  match (u:us) (preq ++ [(p1:ps, rhs), (p2:ps, rhs)] ++ postq) def
```
All I am saying: try out both the current implementation you have, and my suggestion above. I believe that your approach is just as expensive (or more) when it comes to the size of the generated code.

Reply:
So for our small examples it is indeed better to do simple flattening than use let-bindings. However I think if the rhs is large, that it could still be beneficial to wrap (only the rhs) in a let, instead of the rhs + remaining patterns. So I think doing Option 2 is still worth doing, however it will require lots of "extra" implementation to extract free term variables.

Reply:
Actually, in some examples the current code is still much better than straight out flattening.
```
data Number = Zero | Succ Number
data Tuple (a :: *) (b :: *) (c :: *) (d :: *) = Double a b | Triple a b c | Quad a b c d
data List (a :: *) = Nil | Cons a (List a)

-- Test has guard patterns and includes an OR pattern
\x. case x of
    Cons x _ || Cons _ (Cons x Nil)
      | Double _ a <- x, Double b _   <- a -> Succ b
    Nil -> Zero
```
Produces code size (length of string):
Flattening: 7444
Let-bindings: 1374

Reply:
I have implemented "Option 2" from this comment (https://github.com/reilem/KHC/pull/27#discussion_r407426475) and it's even more optimised now. This code now compiles to following lengths (string length without whitespaces):

```
data Tuple (a :: *) (b :: *) (c :: *) (d :: *) = Double a b | Triple a b c | Quad a b c d
data Number = Zero | Succ Number

-- Tests a nested usage of OR patterns. Should succeed.
\i. case i of
  Double (Triple x _ _ || Quad x _ _ _) (Succ y) -> x
  Double (Double _ x || Triple _ x _) Zero -> x
```

Inefficient let-binding: 825
Optimised let-binding: 715
Simple flattening: 599

So now we have the best of both worlds. I can still use let bindings to reduce massive right hand sides into manageable outputs. But some of the inefficient effect on small code has been reduced. In small programs however flattening will of course always be best because the addition of let bindings is often many times the character count of the rhs (which in the small example above is just one character).